Ignored catch parameter

tinganho opened the pull request:
<!--
Thank you for submitting a pull request!

Here's a checklist you might find useful.
[ ] There is an associated issue that is labelled
  'Bug' or 'Accepting PRs' or is in the Community milestone
[ ] Code is up-to-date with the `master` branch
[ ] You've successfully run `jake runtests` locally
[ ] You've signed the CLA
[ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Fixes #17467

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -2196,6 +2196,10 @@
         "category": "Error",
         "code": 2713
     },
+    "Duplicate identifier '_ignoredCatchParameter'. Compiler uses the parameter declaration '_ignoredCatchParameter' to bind ignored catched exceptions.": {
```
If you still need this:

```
Compiler uses the declaration '_ignoredCatchBinding` to bind ignored 'catch' clause parameters.
```

tinganho replied:
Removed.

DanielRosenwasser commented on src/compiler/transformers/esnext.ts:
```diff
@@ -212,6 +214,13 @@ namespace ts {
 
+        function visitCatchClause(node: CatchClause): CatchClause {
+            if (!node.variableDeclaration) {
+                return updateCatchClause(node, createVariableDeclaration("_ignoredCatchParameter"), node.block);
```
Instead of `createVariableDeclaration`, we have a `createUniqueName` that you can call, which relieves you of having to write the error out. I think @RyanCavanaugh was just joking about reserving a variable name (but I'm not sure 😄 ).

rbuckton replied:
Yes, you should either use createUniqueName or createTempVariable.

tinganho replied:
Added `createTempVariable(undefined)` here.

DanielRosenwasser commented:
Looks great other than the unique identifier change I suggested (which may lead you to remove some tests relating to reserving these identifiers).

I think @rbuckton should review the changes.

DanielRosenwasser commented:
(Also, thanks for the PR @tinganho, it's always awesome to see you back!)

rbuckton commented on src/compiler/binder.ts:
```diff
@@ -2921,7 +2921,10 @@ namespace ts {
+        if (!node.variableDeclaration) {
+            transformFlags |= TransformFlags.AssertESNext;
+        }
+        else if (/* node.variableDeclaration && */ isBindingPattern(node.variableDeclaration.name)) {
```
The comment is unnecessary

tinganho replied:
Removed.

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -20885,6 +20885,17 @@ namespace ts {
+                else if (/* !catchClause.variableDeclaration && */ languageVersion < ScriptTarget.ESNext) {
+                    const blockLocals = catchClause.block.locals;
+                    if (blockLocals) {
+                        forEachKey(blockLocals, caughtName => {
```
Do we need a name? We can generate a name in the transforms.

tinganho replied:
I removed this block of change. Since we don't need check and error on a named variable anymore.

rbuckton commented on src/compiler/transformers/es2015.ts:
```diff
@@ -3173,7 +3173,7 @@ namespace ts {
             const ancestorFacts = enterSubtree(HierarchyFacts.BlockScopeExcludes, HierarchyFacts.BlockScopeIncludes);
             let updated: CatchClause;
-            if (isBindingPattern(node.variableDeclaration.name)) {
+            if (node.variableDeclaration && isBindingPattern(node.variableDeclaration.name)) {
```
By the time we get to the es2015 transform we **must** already have a variableDeclaration. Filling in the missing binding should be handled in the esnext transform. At best we should have a debug assertion.

DanielRosenwasser replied:
`Debug.assert(!!node.variableDeclaration)` would be ideal.

rbuckton commented on src/compiler/types.ts:
```diff
@@ -1807,8 +1807,8 @@ namespace ts {
         kind: SyntaxKind.CatchClause;
-        parent?: TryStatement;
-        variableDeclaration: VariableDeclaration;
+        parent?: TryStatement; // We parse missing try statements
```
What does this comment mean? This property is present to reduce costs when walking up the AST.

tinganho replied:
It wasn't apparent to me that the `parent` can be missing in a catch clause.

DanielRosenwasser commented:
Sounds like you want `createTempVariable`, not `createUniqueName`. Is that accurate @rbuckton?

rbuckton commented:
Yes, specifically `createTempVariable(undefined)`.

tinganho commented:
(Thanks @DanielRosenwasser , I have been quite busy for a while, so I haven't have time to participate so much here. Though, working with TS projects all day 😄 )

tinganho commented:
I now, removed the named variable `ignoredCatchParameter`, it made more sense to make the suggested temporary unique variable. So I cleaned up diagnostics, tests and some other unnecessary stuff.

weswigham commented on src/compiler/transformers/esnext.ts:
```diff
@@ -212,6 +214,13 @@ namespace ts {
 
+        function visitCatchClause(node: CatchClause): CatchClause {
+            if (!node.variableDeclaration) {
+                return updateCatchClause(node, createVariableDeclaration(createTempVariable(/*recordTempVariable*/ undefined)), node.block);
```
`node.block` still needs to be recursively visited here, no?
```ts
var foo: any;
try {
   throw "whatever";
}
catch {
  try {
     throw "again"
  }
  catch {
    for await (const c of foo) {
      c;
    }
  }
}

rbuckton replied:
@weswigham is correct, you should recursively visit the block of the catch clause.

tinganho commented:
@weswigham I think I know why, it messes up checking with the lines after it:

![screen shot 2017-08-02 at 11 14 51](https://user-images.githubusercontent.com/1324476/28867121-51078f9c-7775-11e7-8132-7fc85d98dc5c.png)

tinganho commented:
I updated the code with recursive visits.

rbuckton commented on src/compiler/emitter.ts:
```diff
@@ -2131,10 +2131,12 @@ namespace ts {
+            if (node.variableDeclaration) {
+                writeToken(SyntaxKind.OpenParenToken, openParenPos);
+                emit(node.variableDeclaration);
+                writeToken(SyntaxKind.CloseParenToken, node.variableDeclaration ? node.variableDeclaration.end : openParenPos);
```
The conditional on this line is no longer required since you are checking `node.variableDeclaration` on [2134](https://github.com/Microsoft/TypeScript/pull/17517/files#diff-20a9fa4d3e9d09f60c275a3dda245867R2134).

tinganho replied:
✔️

rbuckton commented on src/compiler/parser.ts:
```diff
@@ -4778,11 +4778,12 @@ namespace ts {
-            if (parseExpected(SyntaxKind.OpenParenToken)) {
+
+            if (parseOptional(SyntaxKind.OpenParenToken)) {
                 result.variableDeclaration = parseVariableDeclaration();
```
Conditionally setting `result.variableDeclaration` introduces polymorphism as we now have two `CatchClause` shapes, one with the property and one without. This can cause performance degradation in V8 (NodeJS). I would recommend you add an `else` clause that sets `result.variableDeclaration = undefined` to keep the shape of `CatchClause` the same regardless of which form the user writes.

tinganho replied:
✔️

rbuckton commented on src/compiler/transformers/es2015.ts:
```diff
@@ -3173,6 +3173,7 @@ namespace ts {
         function visitCatchClause(node: CatchClause): CatchClause {
             const ancestorFacts = enterSubtree(HierarchyFacts.BlockScopeExcludes, HierarchyFacts.BlockScopeIncludes);
             let updated: CatchClause;
+            Debug.assert(!!node.variableDeclaration, "Catch clauses should always be present when downleveling ES2015 code.");
```
How about `"Catch clause variable should always be present when downleveling ES2015"`?

tinganho replied:
✔️

rbuckton commented on src/compiler/types.ts:
```diff
@@ -1807,8 +1807,8 @@ namespace ts {
         kind: SyntaxKind.CatchClause;
-        parent?: TryStatement;
-        variableDeclaration: VariableDeclaration;
+        parent?: TryStatement; // We make this optional to parse missing try statements
```
The comment is unnecessary. We only re-introduce `parent` here to refine its type for methods in the checker. `parent` is optional because it parent pointers are not always set, especially during tree transformations.

tinganho replied:
✔️

rbuckton commented on src/compiler/types.ts:
```diff
@@ -1807,8 +1807,8 @@ namespace ts {
-        parent?: TryStatement;
-        variableDeclaration: VariableDeclaration;
+        parent?: TryStatement; // We make this optional to parse missing try statements
+        variableDeclaration?: VariableDeclaration;
```
Can you update `createCatchClause` and `updateCatchClause` in _factory.ts_ and add `| undefined` to the `variableDeclaration` parameters in both cases? This is needed for those using the Compiler API with `-strictNullChecks` enabled for their project.

tinganho replied:
✔️

rbuckton commented:
> @weswigham I think I know why, it messes up checking with the lines after it:
>
> ![screen shot 2017-08-02 at 11 14 51](https://user-images.githubusercontent.com/1324476/28867121-51078f9c-7775-11e7-8132-7fc85d98dc5c.png)

_invalidTryStatements.ts_ covers grammar errors handled in the checker. _invalidTryStatements2.ts_ covers parse errors. I think _invalidTryStatements2.ts_ is a better place for `try {} catch () {}`.

tinganho commented:
> invalidTryStatements.ts covers grammar errors handled in the checker. invalidTryStatements2.ts covers parse errors. I think invalidTryStatements2.ts is a better place for try {} catch () {}.

I added both the test cases `try { }` and `try { } catch { }` in invalidTryStatements2.ts.

DanielRosenwasser commented:
🔔

rbuckton commented:
Looks great! Thanks for the contribution!

tinganho commented:
Thanks for the great review!
